Add `sival_rom_ext` tests to Earlgrey regression & update Bazel test runner.

AlexJones0 opened the pull request:
This PR does 3 related things:
 * Now that upstream OpenTitan `earlgrey_1.0.0` has [merged support for the `sim_qemu_sival_rom_ext`](https://github.com/lowRISC/opentitan/pull/28376) execution environment, these tests will also be passing in the Earlgrey regression that we run on QEMU PRs. Update the list of tests to include these new passing tests, and split the CI runs into two separate jobs to keep CI time shorter.
 * Address the suggestion in https://github.com/lowRISC/qemu/pull/200#issuecomment-3351013600, combining the two tests lists into one list with labelled tags, and optional comments added to tests.
 * Resolve an issue where running the `run-bazel-tests.sh` script locally could delete some local OpenTitan development files, as it was designed for CI.


@rivos-eblot one open question here - CI time is now increased to ~1 hour on the Earlgrey regression when running these tests. It is nice to have the extra coverage (e.g. due to SiVal OTP changes and ROM_EXT locking things down, it is good to test that things pass separately in `rom_with_fake_keys` and `sival_rom_ext`) but this is a double in CI time. Is this still acceptable, or do we need to think about adding a mechanism for only running certain tests in certain environments?

AlexJones0 committed:
[ot] scripts/opentitan: tests: Add `_sival_rom_ext` tests

Now that upstream OpenTitan supports the `sim_qemu_sival_rom_ext`
environment for booting into Owner code after ROM_EXT has run (in
`earlgrey_1.0.0`), update the Earlgrey regression test list to contain
all the newly passing tests.

Signed-off-by: Alex Jones <alex.jones@lowrisc.org>

AlexJones0 committed:
[ot] scripts/opentitan: run-bazel-tests.sh: Combine passing & flaky tests

Store both passing and earlgrey flaky tests in one file, marked by a
starting tag. Also strip trailing comments from this file to allow you
to provide extra context for test results and better track these in the
git history.

Signed-off-by: Alex Jones <alex.jones@lowrisc.org>

AlexJones0 committed:
[ot] scripts/opentitan: run-bazel-tests.sh: Don't clean local overrides

Currently, the `run-bazel-tests.sh` script assumes it is being run from
CI and thus cleans up the local Bazel override files it makes. It is
useful to run it as a developer before making a PR however, but the
current logic would both (a) error upon creating these Bazel override
files if you already had them, and (b) delete them afterwards, breaking
existing OpenTitan integration flows. Add conditional checks to this
logic based on whether these files previously existed to avoid this
behaviour.

Signed-off-by: Alex Jones <alex.jones@lowrisc.org>

rivos-eblot commented:
> * Address the suggestion in [[ot] scripts/opentitan: regression: disable failing test #200 (comment)](https://github.com/lowRISC/qemu/pull/200#issuecomment-3351013600), combining the two tests lists into one list with labelled tags, and optional comments added to tests.

Wow, this was fast! :rocket:

I would make another suggestion: it would be nice if we could keep `scripts/opentitan` reserved for scripts and move data to another directory. I do not have any suggestion for the dir path. For the VM configuration, I re-used `docs/config` that was -pre-existing, although it is a bit weird to use `docs` as the top directory. Any suggestion welcome. I also wanted to move OpenOCD config files out of this directory, so that would be a good time to move them as well.

>  Is this still acceptable, or do we need to think about adding a mechanism for only running certain tests in certain environments?

I think it is a bit long, but I do not really have suggestions. Maybe running dedicated runners could help speeding up tests? Do we have some available?

AlexJones0 committed:
[ot] tests/opentitan: tests: Move bazel test list to `tests/opentitan`

`scripts/opentitan` should be reserved for scripts. Instead we can make
a directory in `tests` for opentitan specifically where we can store the
test data.

Signed-off-by: Alex Jones <alex.jones@lowrisc.org>

AlexJones0 commented:
> I would make another suggestion: it would be nice if we could keep scripts/opentitan reserved for scripts and move data to another directory. I do not have any suggestion for the dir path.

That makes sense, I'll have a look at this. Maybe something like `tests/opentitan/data`?

> I think it is a bit long, but I do not really have suggestions. Maybe running dedicated runners could help speeding up tests? Do we have some available?

We do not have dedicated runners available for QEMU CI, and GitHub's public runners should be fairly fast. We do however have a lot of runners available, so maybe actually I can just split this into two separate jobs which will use different runners in parallel? One for `sim_qemu_rom_with_fake_keys` and one for `sim_qemu_sival_rom_ext`

AlexJones0 committed:
[ot] .github/workflows: ci: Separate Earlgrey regression runs by exec env

Extend the `run-bazel-tests.sh` script for running Earlgrey regressions
via Bazel to be able to take an execution environment as an optional
third argument, defaulting to using the "qemu" tag (which covers all
QEMU exec envs), allowing us to retain the same behaviour.

Also extend CI, so now we use two separate runners, to allow us to run a
`sim_qemu_rom_with_fake_keys` and a `sim_qemu_sival_rom_ext` job in
parallel. This should keep CI times shorter despite the increase in
tests.

Signed-off-by: Alex Jones <alex.jones@lowrisc.org>

AlexJones0 commented:
Marking as WIP whilst I experiment with the CI changes.

rivos-eblot commented:
> We do not have dedicated runners available for QEMU CI, and GitHub's public runners should be fairly fast

It may have changed but AFAIR (previous company) public GitHub runners were deadly slow, with only two cores / runners. This may have changed since.

rivos-eblot commented:
> That makes sense, I'll have a look at this. Maybe something like tests/opentitan/data?

Seems fair, yes. Need to find a new home for the remaining configuration files, but this may be part of another PR.

rivos-eblot commented on .github/workflows/ci_regression.yml:
```diff
@@ -0,0 +1,41 @@
+jobs:
+  earlgrey_rom_with_fake_keys:
+    name: ROM
+    uses: ./.github/workflows/eg_regression.yml
```
is the leading `./` required?

AlexJones0 replied:
Yes, as far as I can see from the [docs](https://docs.github.com/en/actions/reference/workflows-and-actions/workflow-syntax#jobsjob_iduses), if you do not put the leading `./` it is interpreted as a reusable workflow from the `.github` owner's `workflows` repo. Maybe GitHub Actions have a special case to handle the missing `./` here but I'd rather stick to the documented syntax.

AlexJones0 committed:
[ot] scripts/opentitan: tests: Remove coverage_test target

This test was removed from upstream OpenTitan `earlgrey_1.0.0` with
changes to test target coverage, and so should no longer be marked as
passing within the QEMU regression list.

Signed-off-by: Alex Jones <alex.jones@lowrisc.org>

AlexJones0 commented:
> It may have changed but AFAIR (previous company) public GitHub runners were deadly slow, with only two cores / runners. 

I think currently public runners are [4 cores with 16 GB memory](https://docs.github.com/en/actions/reference/runners/github-hosted-runners), which should be *okay* so long as we split the jobs.
